app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java

```diff
@@ -0,0 +1,375 @@
+     */
+    private Activity mActivity;
+    /**
+     * The list of entries tracked - each entry will need a geofence.
+     */
+    private LiveData<? extends List<? extends GeofenceEntry>> mLiveData;
+    /**
+     * Provides access to the Geofencing API.
+     */
+    private GeofencingClient mGeofencingClient;
+    /**
+     * Used when requesting to add or remove geofences.
+     */
+    private PendingIntent mGeofencePendingIntent;
+
+    private PendingGeofenceTask mPendingGeofenceTask = PendingGeofenceTask.NONE;
+
+    //=========//
+    // METHODS //
+    //=========//
+
+    public GeofenceManager(Activity activity, LiveData<? extends List<? extends GeofenceEntry>> liveData) {
+        mActivity = activity;
+        mLiveData = liveData;
+        mGeofencingClient = LocationServices.getGeofencingClient(activity);
+    }
+
+    public void refresh() {
+        // Ask for location permissions if not yet granted
+        if (!checkPermissions()) {
+            mPendingGeofenceTask = PendingGeofenceTask.REFRESH;
+            requestPermissions();
+            return;
+        }
+
+        // TODO: remove only irrelevant geofences and add only the new ones
+        // Remove all geofences and back all the relevant ones
+        removeGeofences();
+        addGeofences();
+    }
+
+    /**
+     * Builds and returns a GeofencingRequest. Specifies the list of geofences to be monitored.
+     * Also specifies how the geofence notifications are initially triggered.
+     */
+    private GeofencingRequest getGeofencingRequest() {
+        Log.d(TAG, "Creating and destroying geofences according to alarms list");
+        List<? extends GeofenceEntry> entries = mLiveData.getValue();
+        List<Geofence> geofenceList = new ArrayList<>();
+
+        // Create geofence objects
+        if (entries != null) {
+            for(GeofenceEntry entry : entries)
+            {
+                if (entry.isEnabled()) {
+                    geofenceList.add(new Geofence.Builder()
+                            // Set the request ID of the geofence. This is a string to identify this
+                            // geofence.
+                            .setRequestId(String.valueOf(entry.getId()))
+
+                            .setCircularRegion(
+                                    entry.getLatitude(),
+                                    entry.getLongitude(),
+                                    entry.getRadius()
+                            )
+                            .setExpirationDuration(Geofence.NEVER_EXPIRE)
+                            .setTransitionTypes(Geofence.GEOFENCE_TRANSITION_ENTER)
+                            .build());
+                }
+            }
+        }
+
+        // Specify geofences and initial triggers
+        GeofencingRequest.Builder builder = new GeofencingRequest.Builder();
+        builder.setInitialTrigger(GeofencingRequest.INITIAL_TRIGGER_ENTER);
+        builder.addGeofences(geofenceList);
```

Comment:
What is the diff? Between this add and line 152?

Reply:
This builds the add request that line 152 needs. This is builder's object add function and the other is th function of the client (actually adds according to the request)